### ISSUE TITLE
Could not replicate your paper's result on Validation_Set of AFF-Wild2

### ISSUE THREAD

**Issue by yts-tc** [wtomin/Multitask-Emotion-Recognition-with-Incomplete-Labels#2]
I have tried to replicate your result of AU Detection on your paper (arxiv:2002.03557v1) as below:
1. Downloaded AFF-Wild2 (AU_Set only) and cropped & aligned face images
    on Validation_Set by OpenFace
    (Using face images that "success" value in OpenFace's log is "1") 
2. Downloaded pretrained model
3. run "run_pretrained_model.py" for CNN Model, Calculate (F1 + Acc) / 2

But my result was "0.415 (FYI. F1: 0.251, Acc: 0.579)", it had big difference with yours ("0.626").
Could you tell me how to replicate your result in case of using AFF-Wild2?

**Comment by wtomin** [comment]
I suppose you were trying to replicate the Multitask CNN results on FAU prediction. My result was 0.626, where F1 score was 0.3158, Accuracy was 0.9362.

I'd like to confirm a few things:
1. About the evaluation metric, have you used the "AU_metric" in Multitask-CNN/utils/model_utils.py ?
2. The Best_AU_threshold is very important for the performance. My threshold is [0.1448537 ,0.03918985, 0.13766725, 0.02652811, 0.40589422, 0.15572545,
0.04808964, 0.10848708] for Multitask CNN model. 
3. The Aff-wild2 validation set has around 67200 images.
4. Your results were reported on the ensemble of five student models.

If possible, you could try to use 0.5 as the threshold for every AU, and check the performance of the teacher model and every student model. Though it may vary, the average performance should be around 0.58. If not, there's something wrong with the uploaded models.

**Comment by yts-tc** [comment]
Thank you for your quick response.
I answered your question as below:

1. No. I did not used Multitask-CNN/utils/model_utils.py and built own script.
    But I checked that evaluation metric is same as your "AU_metric".
2. I used same Best_AU_threshold as yours.
3. Yes. Image amounts of Aff-wild2 is 67,173.
4. Yes. I reported "merged" result.
   I understood that this result showed ensemble of five student models.

Now I tried to use 0.5 as Best_AU_threshold for all AUs and checked performance.
Results were:
F1                   : 0.268
Acc                 : 0.713
(F1 + Acc) / 2 : 0.490

**Comment by wtomin** [comment]
Can you give me the average performance of the five student models? As well as the performance of the teacher model.

**Comment by wtomin** [comment]
Please refer to my recently updated script: eval_val_set.py. It makes it easier for you to evaluate my models on the Aff-wild2 validation set.

To run this script, you need to change the setting from line 8 - 11 to your need. Note that the # annotation file is a dictionary contains the label and the path for each frame in the dataset. It is created from create_annotation_file/Aff-wild2/create_train_val_annotation_file.py. For your ease, I will provide you with my [annotation file](https://hkustconnect-my.sharepoint.com/:u:/g/personal/ddeng_connect_ust_hk/EUu3Z32Z0OxLrof5pXRKSmEBFFOU6B4x49Z3apVPPuGpOg?e=NeYqBO).  You can easily run with it. But I have concern that the discrepancy between your results with mine is probably caused by the labels. So I would recommend you to create your own annotation file. Just follow the same naming convention and data structure with my [annotation file](https://hkustconnect-my.sharepoint.com/:u:/g/personal/ddeng_connect_ust_hk/EUu3Z32Z0OxLrof5pXRKSmEBFFOU6B4x49Z3apVPPuGpOg?e=NeYqBO). And then you can compare the results with two annotation files to see if they differ.

The outputs of `eval_val_set.py` on my computer suggest that the models are correct. Hope this script can help you to replicate my results on the validation set of Aff-wild2.
```
For AU, the performance format is: Result (F1 score, Accuracy)
For EXPR, the performance format is: Result (F1 score, Accuracy)
For VA, the performance format is: Result (Valence CCC, Arousal CCC)
100%|█████████████████████████████████████████████| 7/7 [00:01<00:00,  3.99it/s]
Model teacher performance on AU: 0.5770591195479523 (0.20924365453921795, 0.9448745845566866)
Model student_0 performance on AU: 0.5938610334987143 (0.24260529575579337, 0.9451167712416353)
Model student_1 performance on AU: 0.5831466332508626 (0.2190508413560405, 0.9472424251456846)
Model student_2 performance on AU: 0.5795697135221916 (0.21205721770566466, 0.9470822093387186)
Model student_3 performance on AU: 0.5783187210816221 (0.2093279499355738, 0.9473094922276704)
Model student_4 performance on AU: 0.5733174925926972 (0.2004973039179755, 0.9461376812674188)
Model merged performance on AU: 0.6259866059518737 (0.31581175908171366, 0.9361614528220338)
100%|█████████████████████████████████████████████████████████████████████████████████████████████| 71/71 [00:07<00:00,  9.40it/s]
Model teacher performance on VA: 0.782314152384183 (0.32595055418994784, 0.4563635981942352)
Model student_0 performance on VA: 0.8743277714694281 (0.37001907562751324, 0.5043086958419148)
Model student_1 performance on VA: 0.8763747120724873 (0.3698883081353802, 0.5064864039371071)
Model student_2 performance on VA: 0.8415498997471067 (0.3545625877506158, 0.4869873119964909)
Model student_3 performance on VA: 0.853755630192319 (0.3683833981479413, 0.48537223204437774)
Model student_4 performance on VA: 0.8645130585420513 (0.3533549707109745, 0.5111580878310767)
Model merged performance on VA: 0.8856643571976874 (0.3727163185296373, 0.5129480386680502)
100%|█████████████████████████████████████████████████████████████████████████████████████████████| 71/71 [00:20<00:00,  3.43it/s]
Model teacher performance on EXPR: 0.42192997242577374 (0.3670092705448648, 0.5334356398809523)
Model student_0 performance on EXPR: 0.4367339652648641 (0.37482156662666277, 0.5624348958333333)
Model student_1 performance on EXPR: 0.45384347494174154 (0.3925462034933598, 0.5782955109126984)
Model student_2 performance on EXPR: 0.44852837042452104 (0.3868807523216021, 0.5736917162698413)
Model student_3 performance on EXPR: 0.4439792148214271 (0.38072313068585417, 0.5724082341269842)
Model student_4 performance on EXPR: 0.444058377713212 (0.3808382303288878, 0.5724144345238096)
Model merged performance on EXPR: 0.4583634049224841 (0.3941449740314262, 0.5887462797619047)
```

**Comment by yts-tc** [comment]
>Can you give me the average performance of the five student models? As well as the performance of the teacher model.

F1, Acc & (F1 + Acc) / 2 on each models are below:
(All Best_AU_Threshold are 0.5)
student_0 : 0.258 / 0.705 / 0.481
student_1 : 0.282 / 0.701 / 0.491
student_2 : 0.273 / 0.706 / 0.490
student_3 : 0.246 / 0.709 / 0.477
student_4 : 0.256 / 0.714 / 0.486
teacher     : 0.269 / 0.684 / 0.476

>Please refer to my recently updated script: eval_val_set.py. It makes it easier for you to evaluate my models on the Aff-wild2 validation set.

 I'll try it soon, please wait for little longer.

**Comment by yts-tc** [comment]
I tried to test using annotation file you shared again, below are results.
(I used Best_AU_threshold is [0.1448537 ,0.03918985, 0.13766725, 0.02652811, 0.40589422, 0.15572545, 0.04808964, 0.10848708)

Model teacher performance on AU: 0.5386500711228078 (0.26900767015297367, 0.808292472092642)
Model student_0 performance on AU: 0.5441001755250815 (0.25861408042703526, 0.8295862706231277)
Model student_1 performance on AU: 0.5538952507833688 (0.28203636948929756, 0.8257541320774401)
Model student_2 performance on AU: 0.5521691636129962 (0.27347778204513484, 0.8308605451808576)
Model student_3 performance on AU: 0.5396371741551262 (0.24632354574083792, 0.8329508025694145)
Model student_4 performance on AU: 0.5473701878940924 (0.2559137516470161, 0.8388266241411687)
Model merged performance on AU: 0.4710422770728476 (0.24952564689201773, 0.6925589072536775)

At this time, I could not replicate your result.
I'm going to create own annotation file and test again.

By the way, according to your evaluation script (eval_val_set.py), you assume the input image's extension is ".jpg". When I use OpenFace, OpenFace outputs cropping & aligning facial images as ".bmp".
Did you use ".jpg" images for performance evaluation on your papers?

**Comment by wtomin** [comment]
About the image extension, because I used the crop & aligned face images provided by the challenge organizer, they are in '.jpg' format for the training set,  the validation set and the test set.

 Honestly, I don't know whether the same image in '.bmp' and '.jpg' format can lead to different outputs. If so, it will be a flaw that we didn't expect. 

I will take some time to crop faces using OpenFace, and evaluate my models on those images produced by OpenFace. Please wait for my updates.

**Comment by yts-tc** [comment]
Thank you for your kindly support. I'm waiting for your updates.

For your information, I changed image extension from '.bmp' to ''.jpg", I had performance evaluation again. The result was almost same as ".bmp"'s . (mentioned on 3/16 above)

**Comment by wtomin** [comment]
Great, I think the format (jpg, or bmp) should not change the input images significantly.

While I was using OpenFace API: FaceLandmarkVidMulti to detect face region and do face alignment, I found that the OpenFace would sometimes misclassify persons. Because in this dataset, some video contains the annotations for two persons. The according video sample will be named as "xxx_left" or "xxx_right". OpenFace would misclassify the left person as the right person (or even the third person) for some frames. 

Here is an example. The frames have the same naming convention: "frame_det_#personid_#frameid.bmp". You can see that the faces are misclassified by OpenFace.
![Capture](https://user-images.githubusercontent.com/33117903/76932824-cbf92d00-6926-11ea-95f8-e81ee340babe.PNG)

Can you check whether this kind of misclassification exists in your crop&aligned face images?
I can verify that in the face images provided by the challenge organizer, such mistake is very rare (I haven't found one).

**Comment by yts-tc** [comment]
Thank you for your reply. I used "FeatureExtraction" on OpenFace API.
I try to avoid, checked and removed misclassified person by below procedure:

1.  Checked misclassified person on cropped and aligned images by using facial recognition SW.
2.  I painted out of target region black before executing "FeatureExtraction".
     For example, if name of video sample is "xxx_left", I painted right region black.
     Moreover, if I found misclassified person in procedure 1, I painted him black to avoid to detect.
3.  Checked OpenFace's log, Using face images that "success" value in is "1". 
     "0" means OpenFace failed to detect faces.

I also removed AU's labels corresponding with them.

**Comment by wtomin** [comment]
How is your result recently?

**Comment by yts-tc** [comment]
I made ”annotation.pkl” with own data” and had performance evaluation.

Model teacher performance on AU: 0.538915378123206 (0.26891786457095473, 0.8089128916754572)
Model student_0 performance on AU: 0.5438485619928004 (0.2578150166267823, 0.8298821073588185)
Model student_1 performance on AU: 0.5532957352989759 (0.280665118212613, 0.8259263523853388)
Model student_2 performance on AU: 0.5507457234018838 (0.2704430417344583, 0.8310484050693093)
Model student_3 performance on AU: 0.5397542940758783 (0.24629473223780995, 0.8332138559139466)
Model student_4 performance on AU: 0.5475158038810505 (0.25607540706975096, 0.83895620069235)
Model merged performance on AU: 0.47162665463737363 (0.25019946643522095, 0.6930538428395263)

The result seems to be almost same when I use "annotation.pkl" you provide.
I'm waiting for your update.

**Comment by yts-tc** [comment]
Could you tell me your recently result?

**Comment by wtomin** [comment]
I did not continue to do experiments on the face images obtained by OpenFace, because they are not clean data. I recommend you to send an email to dimitrios.kollias@imperial.ac.uk to apply for the clean training dataset.

**Comment by yts-tc** [comment]
OK, I understood your situation. Thank you for your kindly support.
